changes/2023_7_12_update-keystore-structure/background.md

```diff
@@ -0,0 +1,286 @@
+
+No, having this data be unauthenticated, but then add it to an authenticated process is crazy.
+
+### Just use the existing values on the active entry without authentication
+
+No, this is less dangerous than just using unauthenticated data, but it creates an authentication race. Having a way to inject unauthenticated data into an authenticated process is a bad idea.
+
+### Get the values from an authenticated meta entry
+
+The authentication could be done with decrypt, but this would give decrypt to the create/rotation process.
+This could be constrained with policy, but this is not correct by construction.
+If a customer does not apply this policy then the process is more privileged.
+If it uses the same reEncrypt as above, then what additional value does this provide?
+It also creates a new value that is not a branch key that is now floating around.
+
+## How can we let customers control the branch key id?
+
+### _Require that encryption context MUST also be provided when creating with a defined branch key id_
+
+Since the customer is providing the id value, they know what the value is and can therefore construct the requirement around it.
+Even if these values are literally the same, this adds values.
+Since then the unsafe condition is when the customer system has duplicates.
+Since creating a duplicate will fail, if they continue from there we have to assume some level of shared responsibility.
+Since a customer _could_ just use an existing branch key guid.
+
+### An optional branch key id on create
+
+Simplify given the option is possible.
+This makes the feature “easy to use” but it is not “hard to misuse”.
+
+### The option is still not safe
+
+If duplicates are created in a single key store there is a problem.
+Even GUIDs do not 1000% remove this possibility, however the new structure does.
+Even a duplicate GUID would fail in a single key store.
+Given this, a very poor id selection would also fail.
+This leads us to our second problem, distributed duplicates.
+If the entire system is somehow split brain and a duplicate exists on both sides.
+Simple GUIDs does not save this case. There needs to be further information.
+And the encryption context constraints provide these controls.
+
+Finally, several customers have asked for this.
+With at least one taking on implementing this as custom work.
+It is better for us to provide the safest API we can and document the sharp edge.
+As opposed to trying to not support this or talk about it.
+
+# One-Way Doors
+
+- API changes live forever.
+- Changing the database format renders existing keystores invalid
+
+# Security Considerations
+
+This creates a mechanisms
+to enforce access to branch keys.
+Explicit encryption context verification
+is more secure than implicit encryption context verification.
+
+# Impact/Other considerations
+
+Explaining this subtlety of encryption context
+is going to increase the complexity of our documentation.
```

Comment:
Are we working with docs on this change? Not so much the "why" of these decisions, but more the "how" to use them (e.g. providing custom encryption context, the prefix we add to custom values)

Reply:
yes